### PR TITLE
move go CI to github actions

### PR DESCRIPTION
**Pull request by dhiaayachi**
This PR move go checks to a github action workflow and bump go version used in the tests to 1.18.1

### CODE REVIEW

File: .github/workflows/check.yml [hunk 1]
```diff
@@ -0,0 +1,77 @@
+name: Checks
+
+on:
+  pull_request:
+
+env:
+  GO_VERSION: "1.18.1"
+
+# This workflow runs for not-yet-reviewed external contributions and so it
```
Comment (eculver) [inline .github/workflows/check.yml:9]: Thank you for adding this note <3

File: .github/workflows/check.yml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    steps:
+      - name: "Fetch source code"
+        uses: actions/checkout@v2
+
+      - name: Install Go toolchain
+        uses: actions/setup-go@v2
+        with:
+          go-version: ${{ env.GO_VERSION }}
+
+      # NOTE: This cache is shared so the following step must always be
```
Comment (eculver) [inline .github/workflows/check.yml:28]: Interesting.. I hadn't seen this be done in other workflows, but good to know.

File: .github/workflows/check.yml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: "Fetch source code"
+        uses: actions/checkout@v2
+
+      - name: Install Go toolchain
+        uses: actions/setup-go@v2
+        with:
+          go-version: ${{ env.GO_VERSION }}
```
Comment (eculver) [inline .github/workflows/check.yml:56]: I'm curious if we should a) match what is currently in CircleCI (Go 1.16) or b) even better, now that we're using GHAs, use a matrix of versions and test against 1.16, 1.17 and 1.18 all at once?

File: .github/workflows/check.yml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+          path: "~/go/pkg"
+          key: go-mod-${{ hashFiles('go.sum') }}
+          restore-keys: |
+            go-mod-
+      - name: "Unit tests"
+        run: |
+          go test ./...
+      - name: "Race Unit tests"
+        run: |
+          go test -race ./serf/...
```
Comment (eculver) [inline .github/workflows/check.yml:43]: Can we run these in parallel? Or just run all the tests with the race detector?

File: .github/workflows/check.yml [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      - name: "go.mod and go.sum consistency check"
+        run: |
+          go mod tidy
+          if [[ -n "$(git status --porcelain)" ]]; then
+            echo >&2 "ERROR: go.mod/go.sum are not up-to-date. Run 'go mod tidy' and then commit the updated files."
+            exit 1
+          fi
+      - name: "go vet"
+        run: |
+          go vet ./...
```
Comment (eculver) [inline .github/workflows/check.yml:77]: I'm curious if we should preserve the existing message for relaying failures? 
https://github.com/hashicorp/serf/pull/648/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47L24

**Review by eculver — COMMENTED** [review]
SOLID. Thanks for doing this. I just have a few small questions.

File: .github/workflows/check.yml (hunk 4 above)
Comment (dhiaayachi) [reply .github/workflows/check.yml:43]: This is not possible the other packages have some races (mostly in test) and we need to fix those before being able to run all the tests with race detector. Yes we can run them in parallel, I will make the change.

File: .github/workflows/check.yml (hunk 5 above)
Comment (dhiaayachi) [reply .github/workflows/check.yml:77]: we are actually missing the gofmt check, I will add it.

File: serf/snapshot_test.go [hunk 6]
```diff
@@ -391,6 +391,7 @@ func TestSnapshotter_leave_rejoin(t *testing.T) {
 }
 
 func TestSnapshotter_slowDiskNotBlockingEventCh(t *testing.T) {
+	t.Skip("Flaky test")
```
Comment (dhiaayachi) [inline serf/snapshot_test.go:394]: @banks This is a test you wrote a while back. It's constantly flaking and I'm wondering if it's a perf regression or an issue with the test environment (it flake in circle as much as GHA). 

In any case, I think this is more a benchmark test and we will probably need to write it using a different approach. 
What do you think? is it safe to disable it for the time being (or even remove it) or this is a regression that we need to spend more time investigating?

File: .github/workflows/check.yml [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+          go-version: ${{ matrix.GO_VERSION }}
+
+      # NOTE: This cache is shared so the following step must always be
+      # identical across the unit-tests and consistency-checks
+      # jobs, or else weird things could happen.
+      - name: Cache Go modules
+        uses: actions/cache@v3
+        with:
+          path: "~/go/pkg"
+          key: go-mod-${{ hashFiles('go.sum') }}
```
Comment (eculver) [inline .github/workflows/check.yml:93]: Do we need to bake the `${{ matrix.GO_VERSION }}` into our cache key? I'm curious if the mod cache is version-specific 🤔.

Comment (dhiaayachi) [reply .github/workflows/check.yml:93]: good question, I need to check this. I would say yes we need to.

File: serf/snapshot_test.go (hunk 6 above)
Comment (banks) [reply serf/snapshot_test.go:394]: 👍 

```
        // I typically see this complete in around
	// 115ms on my machine so this should give plenty of headroom for slower CI
	// environments while still being low enough that actual disk IO would
	// reliably blow it.
```

:D I guess I was wrong then! I think it was a more than just a benchmark in the sense that I wanted to verify that our implementation actually didn't allow disk IO to block gossip and effectively break the cluster but yeah see the commentary in https://github.com/hashicorp/serf/pull/526 - we assumed this was lenient enough for CI but if it's not then disabling seems OK?

File: .github/workflows/check.yml (hunk 7 above)
Comment (dhiaayachi) [reply .github/workflows/check.yml]: Added the Go version to the cache key, should be better now.

**Review by eculver — APPROVED** [review]
LGTM

**Pull request merged**